src/main/java/io/cdap/plugin/MongoDBUtil.java

```diff
@@ -0,0 +1,150 @@
+    switch (type) {
+      case RECORD:
+        validateRecordSchema(fieldName, schema, supportedLogicalTypes, supportedTypes);
+        break;
+      case ARRAY:
+        validateArraySchema(fieldName, schema, supportedLogicalTypes, supportedTypes);
+        break;
+      case MAP:
+        validateMapSchema(fieldName, schema, supportedLogicalTypes, supportedTypes);
+        break;
+      default:
+        if (!isSchemaTypeSupported(nonNullableSchema, supportedLogicalTypes, supportedTypes)) {
+          String actualTypeName = logicalType != null ? logicalType.name().toLowerCase() : type.name().toLowerCase();
+          throw new IllegalArgumentException(
+            String.format("Field '%s' is of unsupported type '%s'. Supported types are: %s.",
+                          fieldName, actualTypeName, supportedTypeNames));
+        }
+    }
+  }
+
+  private static void validateMapSchema(String fieldName,
+                                        Schema schema,
+                                        Set<Schema.LogicalType> supportedLogicalTypes,
+                                        Set<Schema.Type> supportedTypes) {
+    Schema nonNullableSchema = schema.isNullable() ? schema.getNonNullable() : schema;
+    Schema keySchema = nonNullableSchema.getMapSchema().getKey();
+    if (keySchema.isNullable()) {
+      throw new IllegalArgumentException(String.format("Key schema can not be nullable for map field '%s'", fieldName));
+    }
+
+    Schema.LogicalType keyLogicalType = keySchema.isNullable() ? keySchema.getNonNullable().getLogicalType()
+      : keySchema.getLogicalType();
+    Schema.Type keyType = keySchema.isNullable() ? keySchema.getNonNullable().getType() : keySchema.getType();
+    if (keyType != Schema.Type.STRING) {
+      String actualTypeName = keyLogicalType != null ? keyLogicalType.name().toLowerCase()
+        : keyType.name().toLowerCase();
+      throw new IllegalArgumentException(String.format("Key schema must be of type 'string', but was '%s' for map " +
+                                                         "field '%s'.", actualTypeName, fieldName));
+    }
+    validateFieldSchema(fieldName, nonNullableSchema.getMapSchema().getValue(), supportedLogicalTypes, supportedTypes);
+  }
+
+  private static void validateArraySchema(String fieldName,
+                                          Schema schema,
+                                          Set<Schema.LogicalType> supportedLogicalTypes,
+                                          Set<Schema.Type> supportedTypes) {
+    Schema nonNullableSchema = schema.isNullable() ? schema.getNonNullable() : schema;
+    Schema componentSchema = nonNullableSchema.getComponentSchema().isNullable()
+      ? nonNullableSchema.getComponentSchema().getNonNullable() : nonNullableSchema.getComponentSchema();
+    validateFieldSchema(fieldName, componentSchema, supportedLogicalTypes, supportedTypes);
+  }
+
+  private static boolean isSchemaTypeSupported(Schema fieldSchema,
```

Comment:
it would be better to make this a validate method that throws an exception like the other methods. Otherwise this isn't really useful because it can be replaced with a single line `return supportedTypes.contains(...) || supportLogicalTypes.contains(...)`

Reply:
Changed to:
```
  private void validateSchemaType(String fieldName, Schema fieldSchema, Set<Schema.LogicalType> supportedLogicalTypes,
                                  Set<Schema.Type> supportedTypes) {
    Schema.Type type = fieldSchema.getType();
    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
    if (supportedTypes.contains(type) || supportedLogicalTypes.contains(logicalType)) {
      return;
    }
	
    String supportedTypeNames = Stream.concat(supportedTypes.stream(), supportedLogicalTypes.stream())
      .map(Enum::name)
      .map(String::toLowerCase)
      .collect(Collectors.joining(", "));
	
    String actualTypeName = logicalType != null ? logicalType.name().toLowerCase() : type.name().toLowerCase();
    throw new IllegalArgumentException(String.format("Field '%s' is of unsupported type '%s'. Supported types are: %s.",
                                                     fieldName, actualTypeName, supportedTypeNames));
  }
```